python/tk_desktop/rpc.py

```diff
@@ -268,3 +393,540 @@ def close(self):
+
+    @property
+    def _engine(self):
+        """
+        Shorthand to access the engine we're exposing methods for.
+        """
+        return self.server.listener.engine
+
+    def do_POST(self):
+        """
+        Handle request to execute a function from the listener.
+        """
+        response = None
+        try:
+            (authkey, (is_blocking, func_name, args, kwargs)) = self._read_request()
+            logger.debug("http server calling '%s(%s, %s)'" % (func_name, args, kwargs))
+
+            # Make sure the right credentials were sent.
+            # Pickling through sgtk.util.pickle converts any binary string to string
+            # so the authkey we received needs to be compared against
+            # the str version.
+            if authkey != six.ensure_str(self.server.listener.authkey):
+                raise ValueError("invalid auth key")
+
+            # Make sure the function exists.
+            if func_name not in self._functions:
+                logger.error(
+                    "unknown function call: '%s', expecting one of '%s'"
+                    % (func_name, list(self._functions))
+                )
+                raise ValueError("unknown function call: '%s'" % func_name)
+
+            # grab the function from the function table
+            func = self._functions[func_name]
+
+            # execute the function on the main thread, as it may to GUI work.
+            if is_blocking:
+                response = self._engine.execute_in_main_thread(func, *args, **kwargs)
+            else:
+                self._engine.async_execute_in_main_thread(func, *args, **kwargs)
+                response = None
+
+            logger.debug("http '%s' result: '%s" % (func_name, response))
+        except Exception as e:
+            response = e
+            # if any of the above fails send the exception back to the client
+            logger.exception("got exception during '%s" % func_name)
+        finally:
+            self._send_response(response)
+
+    def log_request(self, code=None, size=None):
+        """
+        Prevent the server from outputting to stdout every single time
+        a query is processed by the server.
+        """
+
+    def _read_request(self):
+        """
+        Read an incoming request and return the data.
+
+        :returns: The data that was sent.
+        """
+        if six.PY3:
+            content_len = int(self.headers.get("content-length", 0))
+        else:
+            content_len = int(self.headers.getheader("content-length", 0))
+
+        body = self.rfile.read(content_len)
+        try:
+            return pickle.loads(body)
+        except Exception:
+            raise RuntimeError("Response could not be unserialized: %s" % body)
+
+    def _send_response(self, body):
+        """
+        Send a response back to the client.
+
+        :param body: Data to send back to the client.
+        """
+        raw_data = pickle.dumps(body)
```

Comment:
Are we sufficiently protecting the code from loading any rogue/malicious content?
I assume that the authentication is enough to ensure that we are talking to who we think we are?